src/dataframerow/utils.jl

```diff
@@ -149,16 +149,36 @@ function row_group_slots(cols::Tuple{CategoricalVector},
         # we could simply copy refs to groups, but the performance gain is negligible,
         # so always sort groups in the order of levels
         refmap = [0; CategoricalArrays.order(col.pool)]
+        seen = fill(false, length(refmap))
         if skipmissing
             refmap .+= 1
+            seen[1] = true
         else
             refmap[1] = ngroups
         end
         @inbounds for i in eachindex(groups)
-            groups[i] = refmap[col.refs[i]+1]
+            j = refmap[col.refs[i]+1]
+            groups[i] = j
+            seen[j] |= true
+        end
+        if !all(seen)
+            if skipmissing # Always keep first group even if empty
+                ngroups = 1
+                start = 2
+            else
+                ngroups = 0
+                start = 1
+            end
+            @inbounds for i in start:length(refmap)
+                ngroups += seen[i]
+                refmap[i] = ngroups
+            end
+            @inbounds for i in eachindex(groups)
+                groups[i] = refmap[groups[i]]
+            end
         end
     end
-    return ngroups, UInt[], Int[], true, true
+    return ngroups, UInt[], Int[], true
```

Comment:
maybe after squashing non-present groups `groups` vector should be resized just to make sure its length is appropriate (this should not be very expensive I think).

Reply:
Actually it seems we must do it as below you do:
```
stops = zeros(Int, ngroups)
@inbounds for g_ix in groups
    stops[g_ix] += 1
end
```

Can you disable this `@inbounds` in tests and do we have test with categorical vector that has an empty level?

Reply:
Also in lower code it is assumed that `length(groups) == ngroups`
(this means that maybe we should run tests of DataFrames.jl with globaly disabled `@inbounds` with `--check-bounds=yes` option)

Reply:
The `groups` vector has one entry for each row, not one for each group. So AFAICT there's no problem here (and we already pass `--check-bounds=yes`).

Reply:
Thanks. Now I understand.